### ISSUE TITLE
Gutter Icons

### ISSUE THREAD

**Issue by ddnexus** [vsch/idea-multimarkdown#584]
Not sure whether it is a real issue or it's just my ignorance :)
The icons for the link, (Navigate to target.)  are using a lot of horizontal space. 
Is there any way to reduce it with some setting? If not, I would suggest to change the UI so it would show a generic icon with the number of icons, and would open a drop-down menu to actually get the links clicked.

**Comment by vsch** [comment]
@ddnexus, I will add it to the todo list. The IDE displays multiple elements containing line markers as separate line markers. The plugin does not collect the line markers per line. 

The IDE does that, so it is not straight forward because the IDE asks for line markers for an element. Whether they are on the same line or not the plugin does not handle that.

**Comment by vsch** [comment]
@ddnexus, I will add an option to disable line markers. It can be done easily and I have considered it because in my files with multiple ref links on a line I get the ridiculous 3 or 4 line markers.

**Comment by altdd** [comment]
I have the same problem. Looking forward to try it ;). Thanks.

**Comment by altdd** [comment]
Wow! That has been fast! Thanks!

I didn't find a new entry in the IDE Gutter Icon settings. It is the "Show Line Marker" option, right? Still, on my editor the empty gutter is still wide, not as before, but it looks like there is a reserved space... I cannot figure out what should appear there.

**Comment by altdd** [comment]
Not sure I understand the problem... My installed version says 2.5.2.10...

**Comment by vsch** [comment]
@ddnexus, the issue was on my system. I thought it affected the build but it only affected my upgrade procedure. I have so many IDEs for testing that upgrading each one manually is not realistic. I have a an upgrade procedure to update all my IDEs to latest plugin version.

Whew!, the build is correct. The setting is in main Markdown settings:

<img width="771" alt="screen shot 2018-04-17 at 9 29 40 am" src="https://user-images.githubusercontent.com/10299086/38872959-93a784a6-4222-11e8-9fbf-6eec4b30ceae.png">

The IDE keeps the gutter size unchanged unless you close and re-open the file.

**Comment by altdd** [comment]
"The IDE keeps the gutter size unchanged unless you close and re-open the file." I know that... I mean after re-opening the file it is narrower, but it still has space for at least one or two icons.

**Comment by vsch** [comment]
@ddnexus, to reclaim that space you need to turn off gutter icons in the IDE, I can only control what the plugin has control over. The rest is out of my hands.

**Comment by vsch** [comment]
@ddnexus, I can see if there is a setting that I can use to turn off gutter icons for Markdown files. I think there is something if memory serves correctly.

**Comment by ddnexus** [comment]
I found why the space is reserved: you can bookmark a line. So I think the current situation is correct: no need to any other settings IMO.

**Comment by ddnexus** [comment]
Just forgot to report a super small glitch: there is no tool-tip for the new options.

**Comment by vsch** [comment]
@ddnexus, yes I was right there is a setting. New EAP released that will turn off the gutter icons for markdown documents when gutter icons in markdown are disabled.

Also changed the label name to `Show Gutter icons in documents` so does not need a tooltip since it is self explanatory.

**Comment by ddnexus** [comment]
You are fffaaasssttt! :)

**Comment by ddnexus** [comment]
BTW, what about 3 possible options:
1. All the icons visible
2. Turning off the multiple icons but leaving the space for the bookmark icon
3. Turning off he gutter space completely

**Comment by vsch** [comment]
@ddnexus, yes it needs a separate option to turn off the gutter space. This will give the 3 options.

I am not fast, just busy. So try to get small stuff out of the way so I can forget about it, in a good way.. ;)

**Comment by ddnexus** [comment]
And, BTW... the collapse icons are partially gone with the new settings. It shows the folding icons only for the ``` blocks

**Comment by vsch** [comment]
@ddnexus, what do you mean? The two are not related other than the editor not showing the icon gutter. 

If you turn off the gutter icons in the IDE instead of the plugin settings do you get the same thing? If yes you should report it to JetBrains, it would be rendering issue under your OS, I don't have it under OS X.

The plugin only turns off the flag in the editor instance that the IDE does automatically for all editors when you turn off the Gutter Icons in the IDE.

**Comment by ddnexus** [comment]
After I switched on and off the IDE Gutter Icons a couple of times I cannot reproduce it again: now it works again. Maybe there is some sort of interference and it doesn't redraw it correctly so I may had to close/re-open the file, and I am not sure I did. BTW, problem gone. Thanks.

While we are writing a lot... :) Is there any way to use Link Map to point to an attached project from another one? I mean: the file that I would like to link/open is in another project (not the same project root) but it is attached, so the files are somehow available to the current editor...

**Comment by vsch** [comment]
@ddnexus, the plugin uses IDE file indices to locate files. So if the IDE does not index the files then the plugin will not see them.

Link map can be used to change the link text so that the markdown content can have a link which does not match GitHub layout, but is re-mapped for link matching so the plugin link resolver sees it as valid GitHub layout. Like a private and public repositories sharing documentation. The links in the docs are to the public repo files. In the private repo I map the links to the public one so the plugin will resolve the links. See https://github.com/vsch/idea-multimarkdown/wiki/Link-Map-Settings

The only way I do it in my projects that need to refer to other project files is to add the other project as a module in the current one. Then only file:// and http:// absolute addressing links are available since the two are seen as sitting in different VCS roots.

**Comment by ddnexus** [comment]
> The only way I do it in my projects that need to refer to other project files is to add the other project as a module in the current one.

What do you mean with "as a module"? It is a repo on its own, not a module. Do you mean git sub-module? If that is what you mean, it has too many downside for me to be a possible solution.

I have it as an attached project, which get indexed but is not part of the same VCS root, so I suppose it cannot be done, right?

**Comment by vsch** [comment]
@ddnexus, took me a while to get used to the fact that most, not all, JetBrains IDEs can include project directories as modules of other projects.

Had a wild time with breaking the IDE initially because I did not code for that scenario.

So if your particular IDE can include modules and offers crate module from existing sources then you can use that to insert a sub-project from another directory into a project.

Once it is a module its files are indexed as part of the project.

**Comment by ddnexus** [comment]
I just attached project C, which is published. I can confirm, the links find the local reference only from project B or C pointing to files in the main project A, but not the other way around. licking on the link open the file in github, but does not open the file locally in the IDE. Is there any setting to control that?

**Comment by vsch** [comment]
@ddnexus, the plugin uses git VCS roots but has some limitations on resolving multi-module projects. The main project sees the sub-module probably because it is under the project. I say probably because right now the link resolution is a hairy ball of mods that were added as I discovered what is possible in the IDE. Every user brings new use cases.

This link resolution with multi-modules in a predictable and reliable way is on the todo list. It is a major rewrite of the link resolution engine and touches on a ton of code.

It also depends on which link format you use. Page relative is limited to files under the same repo or under the project root. http:// and file:// absolute addressing will work for more cases. To change link format hit the completion again, each time rolls to the next format: page relative, repo relative, http:// then file://, after that rolls over back to page relative. I think I got the order right, but completions show the format in the list.

**Comment by ddnexus** [comment]
Thank you for your explanation. Good to know that it is on the todo list.

> The main project sees the sub-module probably because it is under the project. 
Actually the sub-modules (i.e. the attached projects) see the files in the main project.

I am using the http://github... link format, which is supposed to work when the document will be viewable online, and locally by referencing to the actual local file, not the online url (which works only for links to the main-project files).

Thanks again.
